### ISSUE TITLE
can't load tiff files with repeated errors for the total number of frames should be at least 50.

### ISSUE THREAD

**Issue by KnightyOwl** [MouseLand/suite2p#135]
Hi,

I have a tiff imaging file consistent of 500 frames recorded by two-photon imaging. When I loaded the tiff file with GUI add directory to data_path, the ERROR message kept occurred:

>>>ERROR<<<
Traceback (most recent call last):
  File "/home/anaconda3/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/anaconda3/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/anaconda3/lib/python3.7/site-packages/suite2p/__main__.py", line 32, in <module>
    suite2p.run_s2p.run_s2p(ops, db)
  File "/home/anaconda3/lib/python3.7/site-packages/suite2p/run_s2p.py", line 171, in run_s2p
    ops1[ipl] = register.register_binary(ops1[ipl]) # register binary
  File "/home/anaconda3/lib/python3.7/site-packages/suite2p/register.py", line 363, in register_binary
    op = register_binary(op)
  File "/home/anaconda3/lib/python3.7/site-packages/suite2p/register.py", line 379, in register_binary
    raise Exception('the total number of frames should be at least 50 ')
Exception: the total number of frames should be at least 50 
Interrupted by error (not finished)

I also tested with the example data, however, it worked well in this case. Could there be anything wrong with my tiff files, or any other parts of parameters that I could tune?

Thanks!

**Comment by KnightyOwl** [comment]
Could this because my tiff files are unint16 instead int16?

**Comment by marius10p** [comment]
hi, sorry for the slow reply. That error triggers if fewer than 50 frames are detected, so it must mean the number of frames detected is incorrect. I don't think uint16 has to do with it. 

Is this an imageJ tiff by any chance? If not, what program produced it?

**Comment by Cumol** [comment]
I am having the same issue and it showed up only after I updated suite2p recently, so it wasn't there in the version that I was used the past 6 months.

In my case the tiffs are indeed generated by ImageJ (from MetaMorph series data).

Any idea how to fix that?

**Comment by Cumol** [comment]
@carsen-stringer @marius10p any idea how I can revert to my last version installed? I just saw that there are not "releases" in the repository

**Comment by Cumol** [comment]
I went through the last 20 versions available through pip and found that 0.4.5 is the last version where no problems show up and my files run fine. 0.4.6 gives the above error.

I downgraded to 0.4.5 for now.

**Comment by Cumol** [comment]
Update: it seems to work *sometimes*. As in, I have to run a datatset twice. The first time I get the error above, but the second time it works... I am puzzled

**Comment by carsen-stringer** [comment]
o.o have you tried the github version of suite2p? I did have a bug with the number of frames that is only fixed in the github version

does 0.4.5 always work?

sorry for the delay in dealing with this!

can either of you share a tiff by putting it in this folder? (https://drive.google.com/drive/folders/1OE8FCNfRKmWfAUAVT39IGjXHAmSS1EKd?usp=sharing)

**Comment by carsen-stringer** [comment]
great thanks, I will get to this tonight and should be able to isolate the issue thanks to all the testing you've done!

**Comment by Cumol** [comment]
So, just tested the github version (using `git clone `-> `pip install .`) and it gives the above error every time.

Downgrading to 0.4.5

**Comment by KnightyOwl** [comment]
> o.o have you tried the github version of suite2p? I did have a bug with the number of frames that is only fixed in the github version
> 
> does 0.4.5 always work?
> 
> sorry for the delay in dealing with this!
> 
> can either of you share a tiff by putting it in this folder? (https://drive.google.com/drive/folders/1OE8FCNfRKmWfAUAVT39IGjXHAmSS1EKd?usp=sharing)

My follow-up issue is, I did the deinterleave of channel 1  and channel 2 in imagej, and saved the functional channel 1 as tiff, then when I run suite2p with the tiff saved it returned the error. However, if I load in suite2p with both channels tiff file (the original one generated by scanimage), it ran well.

**Comment by cgohlke** [comment]
This is due to an optimization in the old tifffile module shipped with scikit-image: by default only the first page/IFD of ImageJ hyperstack files is used and reported. To disable this, pass `fastij=False` to [TiffFile()](https://github.com/MouseLand/suite2p/blob/master/suite2p/utils.py#L238) and [imread()](https://github.com/MouseLand/suite2p/blob/master/suite2p/utils.py#L249). Btw, the current version of [tifffile](https://pypi.org/project/tifffile/) no longer supports `pages`, `fastij`, or the TiffFile sequence interface.

**Comment by carsen-stringer** [comment]
thanks for the help! what do you recommend as a replacement?

**Comment by carsen-stringer** [comment]
I've found the scanimage tiff reader works for your imageJ hyperstack (and scanimage tiffs of course :))! it's also very fast. I'm going to do some more testing and then add it to the pipeline to replace tifffile

here is the link to the package: https://pypi.org/project/scanimage-tiff-reader/

it seems to be supported on all operating systems as well!

**Comment by KnightyOwl** [comment]
> I've found the scanimage tiff reader works for your imageJ hyperstack (and scanimage tiffs of course :))! it's also very fast. I'm going to do some more testing and then add it to the pipeline to replace tifffile
> 
> here is the link to the package: https://pypi.org/project/scanimage-tiff-reader/
> 
> it seems to be supported on all operating systems as well!

Thanks! Any follow-up update on this issue? I tried to set fastij in Tifffile as False, however didn't solve this issue...

**Comment by cgohlke** [comment]
Works for me:
```diff
diff --git a/suite2p/utils.py b/suite2p/utils.py
index 9aeb8cf..571ece8 100644
--- a/suite2p/utils.py
+++ b/suite2p/utils.py
@@ -235,7 +235,7 @@ def tiff_to_binary(ops):
     which_folder = -1
     for ik, file in enumerate(fs):
         # size of tiff
-        tif = TiffFile(file)
+        tif = TiffFile(file, fastij=False)
         Ltif = len(tif)
         # keep track of the plane identity of the first frame (channel identity is assumed always 0)
         if ops['first_tiffs'][ik]:
@@ -246,7 +246,7 @@ def tiff_to_binary(ops):
             if ix >= Ltif:
                 break
             nfr = min(Ltif - ix, batch_size)
-            im = imread(file, pages = range(ix, ix + nfr))
+            im = imread(file, pages = range(ix, ix + nfr), fastij=False)
             if len(im.shape) < 3:
                 im = np.expand_dims(im, axis=0)
             if type(im[0,0,0]) == np.uint16:

```

**Comment by KnightyOwl** [comment]
> Works for me:
> 
> ```diff
> diff --git a/suite2p/utils.py b/suite2p/utils.py
> index 9aeb8cf..571ece8 100644
> --- a/suite2p/utils.py
> +++ b/suite2p/utils.py
> @@ -235,7 +235,7 @@ def tiff_to_binary(ops):
>      which_folder = -1
>      for ik, file in enumerate(fs):
>          # size of tiff
> -        tif = TiffFile(file)
> +        tif = TiffFile(file, fastij=False)
>          Ltif = len(tif)
>          # keep track of the plane identity of the first frame (channel identity is assumed always 0)
>          if ops['first_tiffs'][ik]:
> @@ -246,7 +246,7 @@ def tiff_to_binary(ops):
>              if ix >= Ltif:
>                  break
>              nfr = min(Ltif - ix, batch_size)
> -            im = imread(file, pages = range(ix, ix + nfr))
> +            im = imread(file, pages = range(ix, ix + nfr), fastij=False)
>              if len(im.shape) < 3:
>                  im = np.expand_dims(im, axis=0)
>              if type(im[0,0,0]) == np.uint16:
> ```

Thank you very much, the issue has been resolved!

**Comment by carsen-stringer** [comment]
hmm this didn't work for me, maybe I have a different version of the package? thanks for the help @cgohlke !

for long-term use I'm going to integrate the scanimage-tiff-reader into the pipeline, but this fix won't happen for ~1 week, so in the meantime please use this

**Comment by Cumol** [comment]
The fix above did not work for me. Not even after upgrading to 0.5.5 and adding the changes manually.

**Comment by carsen-stringer** [comment]
the tiff you shared @Cumol works correctly with the new scanimage tiff reader but I haven't integrated it into the pipeline yet. I will do this in the next few days, sorry for the delay

**Comment by carsen-stringer** [comment]
I've integrated this into the latest version now, sorry for the delay - this works with the tiff you shared with me (both the scanimagetiffreader and the scikit-image one)

pull the latest github version and also 'pip install scanimage-tiff-reader' to install the new reader

this should fix the other issue you were having too (it was a bug that has since been fixed)

**Comment by Cumol** [comment]
Just tried it out, there seems to be a bug. When I add the directory of the test file and start the pipeline (without changing any settings) I get the following:
```

Found 1 tifs
Opening in GUI (can close this window)

```

And then I get "incorrect file, not a stat.npy, chose another one?". As if the pipeline jumped to the end without doing anything.

I uninstalled suite2p just to make that the installed version is not loaded, installed scanimage-tiff-reader and then cloned the repository freshly and ran python -m suite2p in the cloned directory.

Maybe I should do a fresh install (removing the environment etc.) do is it not necessary?

**Comment by carsen-stringer** [comment]
what? that's so weird, it never said 
"ScanImageTiffReader not working for this tiff type, using scikit-image"?

**Comment by carsen-stringer** [comment]
there was a small bug but you should have seen "ScanImageTiffReader not working for this tiff type, using scikit-image" so I'm not sure if the fix will solve your problem. I'd recommend pulling the latest version regardless. A fresh install shouldn't be necessary but you could try restarting? A failure like that could be memory related :/

**Comment by Cumol** [comment]
Restarted everything and tried again using the GUI, same issue.

Moved to the jupyter notebook and tried there (with hope of getting more info on the error), I get the error "kernel died, restarting".

Something is going terribly wrong....

**Comment by carsen-stringer** [comment]
I reduced the batch size so please pull the latest version - what happened before it died?

in a jupyter-notebook can you see if 'scanimage-tiff-reader' works on Mac, that's the only thing I can think of:
```
from ScanImageTiffReader import ScanImageTiffReader
tif = ScanImageTiffReader(filename)
im = tif.data(beg=0, end=500)
tif.close()
```

**Comment by Cumol** [comment]
Tried again, same behavior. GUI jumps directly to the above error. Jupyter notebook kills the kernel after running the code you poster above.

Went to a normal python prompt and ran the above code, here is the output:
```
Python 3.6.4 |Anaconda, Inc.| (default, Jan 16 2018, 18:10:19) 
[GCC 7.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 
>>> from ScanImageTiffReader import ScanImageTiffReader
>>> 
>>> tif = ScanImageTiffReader('7drg-ablated.tif')
>>> im = tif.data(beg=0, end=500)
Segmentation fault (core dumped)
```

I am running the pipeline on an ubuntu machine.

**Comment by carsen-stringer** [comment]
what version of scanimage-tiff-reader do you have? that was happening in the older version, but I'm not having that problem on my ubuntu machine

this is definitely the issue then unfortunately. does it work with the 'example.tif' you shared? because that one is what I was testing

**Comment by Cumol** [comment]
Running `im = tif.data()` without the beg=0 and end = 500 works however.

**Comment by carsen-stringer** [comment]
can you try 'pip install scanimage-tiff-reader --upgrade'?

**Comment by Cumol** [comment]
```
Requirement already up-to-date: scanimage-tiff-reader in /home/generalubuntu/anaconda3/lib/python3.6/site-packages (1.4.1)
Requirement already satisfied, skipping upgrade: numpy in /home/generalubuntu/anaconda3/lib/python3.6/site-packages (from scanimage-tiff-reader) (1.14.0)
```
I seem to be running the latest version...

Could it be that I have an outdated dependency or something?

**Comment by carsen-stringer** [comment]
hmm maybe? not sure - your tiff is at least 500 frames right? tif.shape() tells you the size

**Comment by Cumol** [comment]
ScanImageTiffReader is only reading the first frame of the file. (tif.shape() return a 2D object with the width and height of the image without a third time/frame dimention)

**Comment by carsen-stringer** [comment]
is this tiff the same format as the 'example.tif' you shared? because this tif worked for both me and Nathan (scanimage developer)

**Comment by cgohlke** [comment]
> ScanImageTiffReader is only reading the first frame of the file

It is possible that large ImageJ files are truncated, i.e. only contain one IFD (from the [tifffile](https://pypi.org/project/tifffile/) documentation): 
```
*ImageJ* hyperstacks store all image data, which may exceed 4 GB,
  contiguously after the first IFD. The size of the image data can be
  determined from the ImageDescription of the first IFD. Files > 4 GB contain
  one IFD only. Tifffile can read and write ImageJ hyperstacks.
```

**Comment by cgohlke** [comment]
Btw, I found the benchmarks at https://vidriotech.gitlab.io/scanimagetiffreader-python/#the-scanimagetiffreader-is-fast "outdated". In my testing on Windows with a 16 GB file, the current version of tifffile performs as well as scanimagetiffreader in the uncached case, and comes "close" to scanimagetiffreader in the cached case when using two threads. But then, tifffile is not specialized on ScanImage TIFFs.

**Comment by Cumol** [comment]
> > ScanImageTiffReader is only reading the first frame of the file
> 
> It is possible that large ImageJ files are truncated, i.e. only contain one IFD (from the [tifffile](https://pypi.org/project/tifffile/) documentation):
> 
> ```
> *ImageJ* hyperstacks store all image data, which may exceed 4 GB,
>   contiguously after the first IFD. The size of the image data can be
>   determined from the ImageDescription of the first IFD. Files > 4 GB contain
>   one IFD only. Tifffile can read and write ImageJ hyperstacks.
> ```

This was a good hint. My file is indeed 5Gb large. I took half of the frames and that is running now without the error. tif.shape() is also giving me the right dimensions. Will report if the pipeline finishes running.

**Comment by carsen-stringer** [comment]
Thank you so much for figuring this out @cgohlke ! Glad it's working now

**Comment by Cumol** [comment]
It does not solve the problem, however, when the files are larger than 4Gb (which are all my files)....

Any idea what can be done then?

**Comment by carsen-stringer** [comment]
Can you put a larger tiff in that Google drive folder (https://drive.google.com/folderview?id=1OE8FCNfRKmWfAUAVT39IGjXHAmSS1EKd ) and I'll post an issue on the scanimage gitlab page?

Is it possible for future acquisitions to reduce the number of pages per tiff? Sorry I know this isn't ideal

**Comment by Cumol** [comment]
"you don't have permission to upload to this folder" ...

I could try downsampling to 2 Hz, but I would rather not to. Additionally, I am probably not the only one with file sizes larger than 4Gb

**Comment by carsen-stringer** [comment]
Sorry I changed the settings of the folder.

Yes there are but they aren't imageJ tiffs usually, other tiff types work at sizes larger than that like ones from scanimage and other acquisition software

**Comment by Cumol** [comment]
Uploaded a full experiment of mine. Feel free to use it for your testing your sparse_mode as well (maybe it is useful for you).

Unfortunately, I am stuck to using Metafluor for acquisition right now (I am also measuring temperature in parallel that needs to be synced with the acquisition) and I have only been able to open their weird tiff format using imagej.

If you have an idea how I can avoid ImageJ, please let me know!

**Comment by carsen-stringer** [comment]
Are you getting single tiffs at a time and then making a stack using imageJ?

**Comment by Cumol** [comment]
Correct. But they are not normal tiffs. I can upload a folder with 10 frames of those files if you want.

**Comment by carsen-stringer** [comment]
Yeah can you put a couple of those in too please? Have you tried running suite2p on these tiffs without stacking them?

**Comment by Cumol** [comment]
I havent tried. They are not .tif files, so I wasn't sure if that will work. I can try though.

**Comment by carsen-stringer** [comment]
oh I misunderstood never mind it wouldn't work then I don't think

**Comment by carsen-stringer** [comment]
but please upload them

**Comment by Cumol** [comment]
Uploaded the first 20 frames of that same large .tif file I uploaded before (they are in a separate folder).

When running suite2p in that folder it gives an error that it did not fine any tiff files.

When I check the file type using `file filename`^in the command line, I get the following output:

`73.001: TIFF image data, little-endian, direntries=13, height=520, bps=16, compression=none, PhotometricIntepretation=BlackIsZero, description=<MetaData>, width=696
`

So they are tiff files, but a weird format?

**Comment by carsen-stringer** [comment]
ah because suite2p looks for things ending in tiff. I've modified sparse_mode to work with this data (single tiffs) but for now you'll have to run it in this notebook: https://github.com/MouseLand/suite2p/blob/sparse_mode/jupyter/run_pipeline_tiffs_or_batch.ipynb

You'll need to have more than 50 tiffs in a folder though because otherwise suite2p will autostop

**Comment by carsen-stringer** [comment]
^ see the last block in that notebook - if you're not familiar with notebooks you need to run 'jupyter-notebook' in your environment (but you can run it in any folder) and make the path at the top of the notebook "sys.path.insert('')" be set to the path to the github repo of suite2p.

**Comment by carsen-stringer** [comment]
and you'll just want to run the first and last blocks for your data (you can take the others out)

**Comment by Cumol** [comment]
The pipeline ran well using the above mentioned notebook!!!!1 That saves me the step of going through imagej!

However, there is another problem. The files are not read in natural order. I am getting all jumbled up traces. maybe add natsort to the import?

**Comment by carsen-stringer** [comment]
Yay so glad! Yeah try natsort.natsorted(files). can you attach a picture of your results to the other thread (since I couldn't run it through with just the 20 tiffs)?

**Comment by kylemxm** [comment]
I had the same problem today running a fresh install of Suite2P on Tif stacks larger than 4 GB saved from ImageJ (Exception: the total number of frames should be at least 50). It seems saving from ImageJ as "filename.tif" by going to "File->save as->OME-TIFF->compression type uncompressed" in FIJI solves the problem and Suite2P can handle the data (Found 1 tifs 3927). 

An older version of Suite2P on my previous laptop had no issue with stacks saved by "File->save as->Tiff..." so it seems to be a recent problem. Hope this helps for those who need to go through ImageJ.

**Comment by carsen-stringer** [comment]
yes this has to do with a recent scikit-image update (see far above). Thanks for this tip, I'll add this info to the wiki. Also you should be able to run on non-stacked tiffs in suite2p (they should in theory work out of the box if they end in *.tif / *.tiff (those are the files suite2p looks for)

**Comment by Cumol** [comment]
@carsen-stringer How do I run the above mentioned notebook (with the list of .tif files) in batch?

Before, I was using the batch jupyter notebook and running `db.append{'data path = ...` for every folder I wanted.

How do I add the step with the .tif files? It seems like I need to generate a new file list for every folder I am interested in. Or am I missing something obvious here?

**Comment by carsen-stringer** [comment]
Yeah for each folder you'll have to add the tifs, I would recommend running in a loop
```
for dp in len(data_path):
      files = glob.glob(os.path.join(dp, "*"))
      for n,f in enumerate(files):
           files[n] = os.path.basename(f)
      db = {'data_path': [dp], 'tiff_list': files}
      run_s2p.run_s2p(db=db, ops=ops)
```

**Comment by carsen-stringer** [comment]
I will add support for this with a flag instead of this solution in the next couple days
